integration-tests/tests/src/tests/sync/app.ts

```diff
@@ -0,0 +1,322 @@
+import { TemplateReplacements } from "../../utils/import-app";
+
+const TestObjectSchema: Realm.ObjectSchema = {
+  name: "TestObject",
+  properties: {
+    doubleCol: "double",
+  },
+};
+
+const PersonForSyncSchema: Realm.ObjectSchema = {
+  name: "Person",
+  primaryKey: "_id",
+  properties: {
+    _id: "objectId",
+    age: "int",
+    dogs: "Dog[]",
+    firstName: "string",
+    lastName: "string",
+    realm_id: "string?",
+  },
+};
+
+const DogForSyncSchema: Realm.ObjectSchema = {
+  name: "Dog",
+  primaryKey: "_id",
+  properties: {
+    _id: "objectId",
+    breed: "string?",
+    name: "string",
+    realm_id: "string?",
+  },
+};
+
+interface IPersonForSyncSchema {
+  _id: ObjectId;
+  age: number;
+  dogs: IDogForSyncSchema[];
+  firstName: string;
+  lastName: string;
+  realm_id: string | undefined;
+}
+
+interface IDogForSyncSchema {
+  _id: ObjectId;
+  breed: string | undefined;
+  name: string;
+  realm_id: string | undefined;
+}
+
+describe("App", () => {
+  describe("instantiation", function () {
+    type ConfigContext = {
+      conf: { id: string; baseUrl: string };
+    } & Mocha.Context;
+    afterEach(async () => {
+      Realm.clearTestState();
+    });
+    before(async function (this: ConfigContext) {
+      this.conf = { id: "smurf", baseUrl: "http://localhost:9090" };
+    });
+
+    it("from config", function (this: ConfigContext) {
+      const app = new Realm.App(this.conf);
+      expect(app).instanceOf(Realm.App);
+    });
+
+    it("from string", function (this: ConfigContext) {
+      const app = new Realm.App(this.conf.id);
+      expect(app).instanceOf(Realm.App);
+      expect(app.id).equals(this.conf.id);
+    });
+
+    it("throws on undefined app", function () {
+      //@ts-expect-error creating an app without a config should fail
+      expect(() => new Realm.App()).throws(Error, "Invalid arguments: 1 expected, but 0 supplied.");
+    });
+
+    it("throws on invalid input", function () {
+      //@ts-expect-error creating an app with an invalid config should fail
+      expect(() => new Realm.App(1234)).throws(Error, "Expected either a configuration object or an app id string.");
+    });
+
+    it("throws on invalid baseURL", function () {
+      const invalid_url_conf = { id: "smurf", baseUrl: "http://localhost:9999" };
+      const app = new Realm.App(invalid_url_conf);
+
+      const credentials = Realm.Credentials.anonymous();
+      console.log(credentials);
+      console.log(app.logIn);
+      return new Promise<void>((resolve, reject) => {
+        return app
+          .logIn(credentials)
+          .then((user) => {
+            return reject(`Able to log in with config ${JSON.stringify(invalid_url_conf)}`);
+          })
+          .catch((err) => {
+            expect(err.message).contains(
+              "request to http://localhost:9999/api/client/v2.0/app/smurf/location failed, reason: connect ECONNREFUSED",
+            );
+            return resolve();
+          });
+      });
```

Comment:
This is better expressed using `chai-as-promised`.
You'd be able to find inspiration in [the "flexible sync" tests](https://github.com/realm/realm-js/blob/bindgen/integration-tests/tests/src/tests/sync/flexible.ts#L219-L221):

```suggestion
      await expect(app.logIn(credentials)).to.be.rejectedWith(
        "request to http://localhost:9999/api/client/v2.0/app/smurf/location failed, reason: connect ECONNREFUSED",
      );
```

Reply:
perfect! Felt like they grew out of hand so this makes it a whole lot easier